Content.Server/Objectives/Systems/KillPersonConditionSystem.cs

```diff
@@ -99,6 +105,47 @@ private void OnHeadAssigned(EntityUid uid, PickRandomHeadComponent comp, ref Obj
         _target.SetTarget(uid, _random.Pick(allHeads), target);
     }
 
+    private void OnTraitorKillAssigned(EntityUid uid, PickRandomTraitorComponent comp, ref ObjectiveAssignedEvent args)
+    {
+        if (!TryComp<TargetObjectiveComponent>(uid, out var target))
+        {
+            Debug.Fail($"Missing components for {uid}.");
+            args.Cancelled = true;
+            return;
+        }
+
+        // target already assigned
+        if (target.Target != null)
+        {
+            Debug.Fail($"Target already assigned for {uid}.");
+            args.Cancelled = true;
+            return;
+        }
+
+        var traitors = _traitorRule.GetOtherTraitorMindsAliveAndConnected(args.Mind);
+
+        // No other traitors
+        if (traitors.Count == 0)
+        {
+            args.Cancelled = true;
+            return;
+        }
+
+        var targetTraitorMind = _random.Pick(traitors).Id;
+
+        // Make sure the current traitor doesn't have any targets targeting this traitor already.
```

Comment:
Wouldn't it make more sense to first filter out these cases and then pick from the remaining traitors list?